src/Titanium.Web.Proxy/Network/CertificateManager.cs

```diff
@@ -472,61 +436,69 @@ internal X509Certificate2 CreateCertificate(string certificateName, bool isRootC
         internal async Task<X509Certificate2> CreateCertificateAsync(string certificateName)
         {
             // check in cache first
-            if (certificateCache.TryGetValue(certificateName, out var cached))
+            var item = cachedCertificates.GetOrAdd(certificateName, _ =>
             {
-                cached.LastAccess = DateTime.Now;
-                return cached.Certificate;
-            }
+                var cached = new CachedCertificate();
+                cached.CreationTask = Task.Run(() =>
+                {
+                    var certificate = CreateCertificate(certificateName, false);
+
+                    // see http://www.albahari.com/threading/part4.aspx for the explanation
+                    // why Thread.MemoryBarrier is used here and below
+                    cached.Certificate = certificate;
+                    Thread.MemoryBarrier();
+                    cached.CreationTask = null;
+                    Thread.MemoryBarrier();
+                    return certificate;
+                });
 
-            // handle burst requests with same certificate name
-            // by checking for existing task for same certificate name
-            if (pendingCertificateCreationTasks.TryGetValue(certificateName, out var task))
-            {
-                return await task;
-            }
+                return cached;
+            });
 
-            // run certificate creation task & add it to pending tasks
-            task = Task.Run(() =>
+            item.LastAccess = DateTime.Now;
+
+            if (item.Certificate != null)
```

Comment:
Shouldn't this be just item!=null ?

Reply:
Okay. I refactored it removing locking while also preventing multiple tasks for same certificate.